lib/Socket.js

```diff
@@ -0,0 +1,149 @@
+  }
+
+  connect(address) {
+    if (!url.parse(address).port) {
+      address += ':' + DEFAULT_PORT;
+    }
+
+    if (this._connected) {
+      this._socket.close();
+      this._connected = false;
+    }
+
+    return new Promise((resolve, reject) => {
+      log.info('Attempting to connect to:', address);
+      this._socket = new WebSocket('ws://' + address);
+
+      this._socket.onopen = () => {
+        log.info('Connection opened:', address);
+        this.emit('obs:internal:event', { updateType: 'ConnectionOpened' });
+        this._connected = true;
+        resolve();
+      };
+
+      this._socket.onclose = () => {
+        log.info('Connection closed:', address);
+        this.emit('obs:internal:event', { updateType: 'ConnectionClosed' })
+        this._connected = false;
+      };
+
+      this._socket.onerror = (evt) => {
+        log.error('Connected failed.', evt.code);
+        this._connected = false;
+        reject(evt);
+      };
+
+      this._socket.onmessage = (msg) => {
+        log.debug('[Socket:OnMessage]', msg);
+
+        var data = camelCaseKeys(JSON.parse(msg.data));
+
+        // Emit the message with ID if available, otherwise default to a non-messageId driven event.
+        if (data.messageId) {
+          log.debug('[Socket:Emit]', 'obs:internal:message:id-' + data.messageId, data);
+          this.emit('obs:internal:message:id-' + data.messageId, data);
+        } else {
+          log.debug('[Socket:Emit]', 'obs:internal:event', data);
+          this.emit('obs:internal:event', data);
+        }
+      };
+    });
+  }
+
+  authenticate(args, callback) {
+    args = args || {};
+    args.password = args.password || '';
+    callback = callback || NOP;
+
+    return this.GetAuthRequired()
+      .then((data) => {
+        AUTH = {
+          required: data.authRequired,
+          salt: data.salt,
+          challenge: data.challenge
+        };
+
+        // Return early if authentication is not necessary.
+        if (!AUTH.required) {
+          log.debug('Authentication Not Required.');
+          this.emit('obs:internal:event', { updateType: 'AuthenticationSuccess' });
+          try { callback(null, true); } catch (e) { }
+          return true;
+        }
+
+        var params = {
+          'auth': new AuthHashing(AUTH.salt, AUTH.challenge).hash(args.password)
+        };
+
+        return this.send('Authenticate', params, callback)
+          .then(() => {
+            log.debug('Authentication Success.');
+            this.emit('obs:internal:event', { updateType: 'AuthenticationSuccess' });
+            try { callback(null, true); } catch (e) { }
+            return Promise.resolve();
+          })
+          .catch(() => {
+            log.error('Authentication Failure.');
+            this.emit('obs:internal:event', { updateType: 'AuthenticationFailure' });
+            try { callback(true, null); } catch (e) { }
+            return Promise.reject();
```

Comment:
Hey there!

Ah yes, Promises can be tricky. The good news is that I think they're more straightforward than you're anticipating, and your code is *very* close to working.

In the current commit (which seems to have been made after this review and has some changes), I see the following things to change ([gist example of the changes in question](https://gist.github.com/Lange/2009b165e9df285de3a28ea55004687a)):

1. Change line 120 to just be `return;` (Or `return true;` or whatever value you want to return)
2. Delete line 138
3. Delete Line 132

### Explanation
There are no static `Promise.resolve()` or `Promise.reject()` methods. Those are instance methods. You'd only call `promise.resolve()` or `promise.reject()` if you had a `Promise` instance to reference.

In this case, the resolution or rejection is happening in the `send` method, and are already part of the promise chain. You then are adding your `then` and `catch` handlers to handle what happens when that Promise is `resolve`d` or `reject`ed.

Let me know if you have any other questions! Happy to help out.